ath79: add support for TP-Link CPE610-v1

apcameron opened the pull request:
TP-Link CPE610-v1 is an outdoor wireless CPE for 5 GHz with
one Ethernet port based on Atheros AR9344

Specifications:
 - Based on the same underlying hardware as the TP-Link CPE510
 - Power, LAN0, WLAN green LEDs
 - 23dBi high-gain directional 2×2 MIMO antenna and a dedicated metal reflector

Flashing instructions:
 Flash factory image through stock firmware WEB UI
 or through TFTP
 To get to TFTP recovery just hold reset button while powering on for
 around 4-5 seconds and release.
 Rename factory image to recovery.bin
 Stock TFTP server IP:192.168.0.100
 Stock device TFTP adress:192.168.0.254

Signed-off-by: Andrew Cameron <apcameron@softhome.net>

apcameron committed:
ath79: add support for TP-Link CPE610-v1

TP-Link CPE610-v1 is an outdoor wireless CPE for 5 GHz with
one Ethernet port based on Atheros AR9344

Specifications:
 - Based on the same underlying hardware as the TP-Link CPE510
 - Power, LAN, WLAN5G green LEDs
 - 23dBi high-gain directional 2×2 MIMO antenna and a dedicated metal reflector

Flashing instructions:
 Flash factory image through stock firmware WEB UI
 or through TFTP
 To get to TFTP recovery just hold reset button while powering on for
 around 4-5 seconds and release.
 Rename factory image to recovery.bin
 Stock TFTP server IP:192.168.0.100
 Stock device TFTP adress:192.168.0.254

Signed-off-by: Andrew Cameron <apcameron@softhome.net>

981213 reviewed (changes requested):
First of all, would you mind to try if Device/tplink-safeloader-uimage in image/generic-tp-link.mk works on your router? (you'll need to change partition layout in tplink-safeloader.c and dts accordingly.) Your current kernel size isn't big enough and will easily get oversized in the future. If the uimage template works we can use a dynamic kernel/rootfs splitting on this router.

Several other comments inline:

dengqf6 commented on target/linux/ath79/base-files/etc/board.d/01_leds:
```diff
@@ -90,6 +90,10 @@ tplink,archer-c6-v2)
 	;;
+tplink,cpe610-v1)
+	ucidef_set_led_netdev "lan" "LAN" "tp-link:green:lan0" "eth0"
+	ucidef_set_led_wlan "wlan" "WLAN" "tp-link:green:link1" "wlan0"
```
@981213 Can be also moved into DTS

981213 replied:
> @981213 Can be also moved into DTS

link1 led is used as system boot indicator so the trigger set in dts will be overrided by diag.sh. That's the issue mentioned in #1698 .

apcameron commented:
@981213 @dengqf6 @p-wassi  It is my hope that PR #1704 can be refactored to take into account all the TP-Link CPE Devices based on the Atheros AR9344 and the Qualcomm Atheros QCA953x SOC's. It would be great if it could provide the base .dtsi files and image configuration standards that can be used by all the TP-Link CPE devices as they are almost identical except for small differences.
Using the OPENWRT_COMMIT=v18.06.1 and in some cases private patches listed below 
I am aware of the following devices being used with OpenWrt around the world based
on the ar71xx branch that need to be taken into account on the ath79 branch

TPLink CPE210 v1.0
TPLink CPE210 v1.1
TPLink CPE210 v2.0
TPLink CPE210 v3.0
TPLink CPE220 v1.0
TPLink CPE220 v2.0
TPLink CPE220 v3.0
TPLink CPE510 v1.0
TPLink CPE510 v1.1
TPLink CPE510 v2.0
TPLink CPE510 v3.0
TPLink CPE520 v1.0
TPLink CPE520 v2.0
TPLink CPE610 v1.0

Some of those patches that are applied to OPENWRT_COMMIT=v18.06.1 are located here.
https://github.com/aredn/aredn_ar71xx/blob/develop/patches/001-add_support_for_TP-Link_CPE210_v3.patch
https://github.com/aredn/aredn_ar71xx/blob/develop/patches/001-add_support_for_TP-Link_CPE220_v2.patch
https://github.com/aredn/aredn_ar71xx/blob/develop/patches/001-add_support_for_TP-Link_CPE220_v3.patch
https://github.com/aredn/aredn_ar71xx/blob/develop/patches/001-add_support_for_TP-Link_CPE510_v2.patch
https://github.com/aredn/aredn_ar71xx/blob/develop/patches/001-add_support_for_TP-Link_CPE510_v3.patch
https://github.com/aredn/aredn_ar71xx/blob/develop/patches/001-add_support_for_TP-Link_CPE610_v1.patch

As such once the standards have been set then will be the time to update my PR unless this device gets included in the refactored PR #1704

apcameron commented on target/linux/ath79/dts/ar9344_tplink_cpe610-v1.dts:
```diff
@@ -0,0 +1,161 @@
+	status = "okay";
+};
+
+&gpio {
```
See comment at the end

ynezz replied:
What comment in the end? Please don't mark the comments as resolved until resolved via git, thanks.

apcameron commented:
@ynezz What should I Change?

ynezz commented:
> What should I Change?

@apcameron I've quickly skimmed through this PR and I've seen merge conflicts in two files and your references to #1704 in regards to common DTSI, dynamic partitions etc.

apcameron commented:
Updated to resolve Merge Conflicts

adrianschmutzler commented on tools/firmware-utils/src/tplink-safeloader.c:
```diff
@@ -244,6 +244,45 @@ static struct device_info boards[] = {
 	},
+	/** Firmware layout for the CPE610 */
+	{
+		.id     = "CPE610",
```
Any chance of having this changed to CPE610V1? If yes, same would be necessary in generic-tp-link.mk.

apcameron replied:
I will upload an update with your suggestions after testing it.

adrianschmutzler commented on target/linux/ath79/image/generic-tp-link.mk:
```diff
@@ -77,6 +77,22 @@ define Device/tplink_archer-c7-v5
+  COMPILE := loader-$(1).elf
+  COMPILE/loader-$(1).elf := loader-okli-compile
+  KERNEL := kernel-bin | append-dtb | lzma | uImage lzma -M 0x4f4b4c49 | loader-okli $(1) 12288
+  SUPPORTED_DEVICES += cpe610-v1
```
SUPPORTED_DEVICES is still there?

ynezz replied:
I'm wondering why are you marking those comments as resolved, this should be done automatically when you `git push -f` and not manually.

adrianschmutzler commented:
> It is my hope that PR #1704 can be refactored to take into account all the TP-Link CPE Devices based on the Atheros AR9344 and the Qualcomm Atheros QCA953x SOC's.

If I'm not mistaken, you have slightly different partition layout and no eth1 for CPE210v2/v3. Don't know how the other devices deviate. Is there enough similarity between all devices to justify a common dtsi?

apcameron commented:
@adrianschmutzler Yes the devices do differ but as more of them are added I think we may find that there will be some configurations/settings that they share in common. However we will not know that until we have a number of the TP-LINK CPE devices on the ATH79 Branch.

apcameron commented:
Here is a copy of the boot log
----
[    0.000000] Linux version 4.14.108 (apcameron@debian) (gcc version 7.4.0 (OpenWrt GCC 7.4.0 r9669+71-94ffb7be4d)) #0 Wed Mar 27 21:40:39 2019
[    0.000000] bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 0001974c (MIPS 74Kc)
[    0.000000] MIPS: machine is TP-Link CPE610 v1
[    0.000000] SoC: Atheros AR9344 rev 2
[    0.000000] Determined physical RAM map:
[    0.000000]  memory: 04000000 @ 00000000 (usable)
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] Primary instruction cache 64kB, VIPT, 4-way, linesize 32 bytes.
[    0.000000] Primary data cache 32kB, 4-way, VIPT, cache aliases, linesize 32 bytes
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] On node 0 totalpages: 16384
[    0.000000] free_area_init_node: node 0, pgdat 80519600, node_mem_map 810078a0
[    0.000000]   Normal zone: 128 pages used for memmap
[    0.000000]   Normal zone: 0 pages reserved
[    0.000000]   Normal zone: 16384 pages, LIFO batch:3
[    0.000000] random: get_random_bytes called from start_kernel+0x8c/0x474 with crng_init=0
[    0.000000] pcpu-alloc: s0 r0 d32768 u32768 alloc=1*32768
[    0.000000] pcpu-alloc: [0] 0 
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 16256
[    0.000000] Kernel command line: console=ttyS0,115200 rootfstype=squashfs,jffs2
[    0.000000] PID hash table entries: 256 (order: -2, 1024 bytes)
[    0.000000] Dentry cache hash table entries: 8192 (order: 3, 32768 bytes)
[    0.000000] Inode-cache hash table entries: 4096 (order: 2, 16384 bytes)
[    0.000000] Writing ErrCtl register=00000000
[    0.000000] Readback ErrCtl register=00000000
[    0.000000] Memory: 58212K/65536K available (3776K kernel code, 151K rwdata, 912K rodata, 1232K init, 204K bss, 7324K reserved, 0K cma-reserved)
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS: 51
[    0.000000] CPU clock: 560.000 MHz
[    0.000000] clocksource: MIPS: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 6825930166 ns
[    0.000010] sched_clock: 32 bits at 280MHz, resolution 3ns, wraps every 7669584382ns
[    0.009077] Calibrating delay loop... 278.93 BogoMIPS (lpj=1394688)
[    0.086304] pid_max: default: 32768 minimum: 301
[    0.091886] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.099580] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.112483] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.123963] futex hash table entries: 256 (order: -1, 3072 bytes)
[    0.131136] pinctrl core: initialized pinctrl subsystem
[    0.138332] NET: Registered protocol family 16
[    0.174131] clocksource: Switched to clocksource MIPS
[    0.181233] NET: Registered protocol family 2
[    0.187201] TCP established hash table entries: 1024 (order: 0, 4096 bytes)
[    0.195367] TCP bind hash table entries: 1024 (order: 0, 4096 bytes)
[    0.202762] TCP: Hash tables configured (established 1024 bind 1024)
[    0.210320] UDP hash table entries: 256 (order: 0, 4096 bytes)
[    0.217164] UDP-Lite hash table entries: 256 (order: 0, 4096 bytes)
[    0.224773] NET: Registered protocol family 1
[    0.229863] PCI: CLS 0 bytes, default 32
[    0.233728] Crashlog allocated RAM at address 0x3f00000
[    0.241449] workingset: timestamp_bits=30 max_order=14 bucket_order=0
[    0.254565] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.261337] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.283928] io scheduler noop registered
[    0.288544] io scheduler deadline registered (default)
[    0.296916] pinctrl-single 1804002c.pinmux: 544 pins at pa b804002c size 68
[    0.306200] Serial: 8250/16550 driver, 1 ports, IRQ sharing disabled
[    0.314459] console [ttyS0] disabled
[    0.318651] 18020000.uart: ttyS0 at MMIO 0x18020000 (irq = 8, base_baud = 2500000) is a 16550A
[    0.328709] console [ttyS0] enabled
[    0.336140] bootconsole [early0] disabled
[    0.353716] m25p80 spi0.0: gd25q64 (8192 Kbytes)
[    0.358539] 6 fixed-partitions partitions found on MTD device spi0.0
[    0.365008] Creating 6 MTD partitions on "spi0.0":
[    0.369885] 0x000000000000-0x000000020000 : "u-boot"
[    0.375913] 0x000000020000-0x000000030000 : "partition-table"
[    0.382619] 0x000000030000-0x000000040000 : "product-info"
[    0.389133] 0x000000040000-0x0000007c0000 : "firmware"
[    0.395279] 2 fixed-partitions partitions found on MTD device firmware
[    0.401915] Creating 2 MTD partitions on "firmware":
[    0.407021] 0x000000000000-0x0000001c0000 : "kernel"
[    0.412946] 0x0000001c0000-0x000000780000 : "rootfs"
[    0.418892] mtd: device 5 (rootfs) set to be root filesystem
[    0.424749] 1 squashfs-split partitions found on MTD device rootfs
[    0.431038] 0x0000003b0000-0x000000780000 : "rootfs_data"
[    0.437399] 0x0000007c0000-0x0000007f0000 : "config"
[    0.443283] 0x0000007f0000-0x000000800000 : "art"
[    0.450376] libphy: Fixed MDIO Bus: probed
[    0.794553] libphy: ag71xx_mdio: probed
[    0.799601] libphy: ar8xxx-mdio: probed
[    0.814032] switch0: Atheros AR8229 rev. 1 switch registered on mdio-bus.0
[    1.196133] ag71xx 19000000.eth: connected to PHY at mdio-bus.0:1f:04 [uid=004dd042, driver=Generic PHY]
[    1.206504] eth0: Atheros AG71xx at 0xb9000000, irq 4, mode: mii
[    1.546118] ag71xx 1a000000.eth: connected to PHY at mdio-bus.0:1f:03 [uid=004dd042, driver=Generic PHY]
[    1.556493] eth1: Atheros AG71xx at 0xba000000, irq 5, mode: gmii
[    1.565075] NET: Registered protocol family 10
[    1.574792] Segment Routing with IPv6
[    1.578633] NET: Registered protocol family 17
[    1.583190] 8021q: 802.1Q VLAN Support v1.8
[    1.593363] VFS: Mounted root (squashfs filesystem) readonly on device 31:5.
[    1.609384] Freeing unused kernel memory: 1232K
[    1.613988] This architecture does not have kernel memory protection.
[    2.267258] init: Console is alive
[    2.270966] init: - watchdog -
[    2.495850] random: fast init done
[    3.144444] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[    3.261088] kmodloader: done loading kernel modules from /etc/modules-boot.d/*
[    3.279192] init: - preinit -
[    4.199174] random: jshn: uninitialized urandom read (4 bytes read)
[    4.375116] random: jshn: uninitialized urandom read (4 bytes read)
[    4.436974] random: jshn: uninitialized urandom read (4 bytes read)
[    4.807739] IPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready
[    6.906734] eth0: link up (100Mbps/Full duplex)
[    6.911374] IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready
[    8.066118] mount_root: jffs2 not ready yet, using temporary tmpfs overlay
[    8.100758] urandom-seed: Seed file not found (/etc/urandom.seed)
[    8.294699] eth0: link down
[    8.316380] procd: - early -
[    8.319429] procd: - watchdog -
[    8.959423] procd: - watchdog -
[    8.962947] procd: - ubus -
[    8.994361] urandom_read: 5 callbacks suppressed
[    8.994370] random: ubusd: uninitialized urandom read (4 bytes read)
[    9.016129] random: ubusd: uninitialized urandom read (4 bytes read)
[    9.024255] procd: - init -
[    9.474104] kmodloader: loading kernel modules from /etc/modules.d/*
[    9.488621] ip6_tables: (C) 2000-2006 Netfilter Core Team
[    9.504807] Loading modules backported from Linux version v4.19.23-0-g67d52fae61c1
[    9.512500] Backport generated by backports.git v4.19.23-1-0-g480a925a
[    9.522785] ip_tables: (C) 2000-2006 Netfilter Core Team
[    9.539042] nf_conntrack version 0.5.0 (1024 buckets, 4096 max)
[    9.615319] xt_time: kernel timezone is -0000
[    9.689770] PPP generic driver version 2.4.2
[    9.697199] NET: Registered protocol family 24
[    9.754280] ath: EEPROM regdomain: 0x0
[    9.754291] ath: EEPROM indicates default country code should be used
[    9.754295] ath: doing EEPROM country->regdmn map search
[    9.754313] ath: country maps to regdmn code: 0x3a
[    9.754320] ath: Country alpha2 being used: US
[    9.754324] ath: Regpair used: 0x3a
[    9.770180] ieee80211 phy0: Selected rate control algorithm 'minstrel_ht'
[    9.772274] ieee80211 phy0: Atheros AR9340 Rev:2 mem=0xb8100000, irq=11
[    9.884643] kmodloader: done loading kernel modules from /etc/modules.d/*
[   74.655641] jffs2_scan_eraseblock(): End of filesystem marker found at 0x0
[   74.662635] jffs2_build_filesystem(): unlocking the mtd device... 
[   74.683684] done.
[   74.691981] jffs2_build_filesystem(): erasing all blocks after the end marker... 
[   79.167942] br-lan: port 1(eth0) entered blocking state
[   79.180907] br-lan: port 1(eth0) entered disabled state
[   79.186586] device eth0 entered promiscuous mode
[   79.254652] IPv6: ADDRCONF(NETDEV_UP): br-lan: link is not ready
[   81.226703] eth0: link up (100Mbps/Full duplex)
[   81.231362] br-lan: port 1(eth0) entered blocking state
[   81.236712] br-lan: port 1(eth0) entered forwarding state
[   81.275563] IPv6: ADDRCONF(NETDEV_CHANGE): br-lan: link becomes ready
[   81.438043] done.
[   81.440044] jffs2: notice: (996) jffs2_build_xattr_subsystem: complete building xattr subsystem, 0 of xdatum (0 unchecked, 0 orphan) and 0 of xref (0 dead, 0 orphan) found.
[   81.703666] overlayfs: upper fs does not support tmpfile.
---

adrianschmutzler commented:
> Yes the devices do differ but as more of them are added I think we may find that there will be some configurations/settings that they share in common. However we will not know that until we have a number of the TP-LINK CPE devices on the ATH79 Branch.

Yes, I also think, at the moment, its better to first add devices and then merge into reasonable DTSIs later.

ynezz commented on target/linux/ath79/dts/ar9344_tplink_cpe610-v1.dts:
```diff
@@ -0,0 +1,161 @@
+	};
+};
+
+&eth1 {
```
I'm wondering why are you marking those comments as resolved, this should be done automatically when you git push -f and not manually.

adrianschmutzler replied:
Is the status = "okay" required here?

ynezz commented on target/linux/ath79/dts/ar9344_tplink_cpe610-v1.dts:
```diff
@@ -0,0 +1,161 @@
+	phy-handle = <&swphy4>;
+};
+
+&builtin_switch {
```
Please don't mark the comments as resolved until resolved via git, thanks.

apcameron commented:
Hi @ynezz,

 

This process is new to me. I will not do that again.

 

I will re-look at those and see what I can change without breaking the device.

 

 

From: Petr Štetiar [mailto:notifications@github.com] 
Sent: Monday, April 01, 2019 10:29 AM
To: openwrt/openwrt <openwrt@noreply.github.com>
Cc: apcameron <apcameron@softhome.net>; Mention <mention@noreply.github.com>
Subject: Re: [openwrt/openwrt] ath79: add support for TP-Link CPE610-v1 (#1748)

 

@ynezz commented on this pull request.

  _____  

In target/linux/ath79/dts/ar9344_tplink_cpe610-v1.dts <https://github.com/openwrt/openwrt/pull/1748#discussion_r270924688> :

> +     status = "okay";
+
+       mtd-cal-data = <&art 0x1000>;
+       mtd-mac-address = <&productinfo 0x08>;
+};
+
+&eth0 {
+       status = "okay";
+
+       mtd-mac-address = <&productinfo 0x08>;
+
+       phy-mode = "mii";
+       phy-handle = <&swphy4>;
+};
+
+&builtin_switch {

Please don't mark the comments as resolved until resolved via git, thanks.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/openwrt/openwrt/pull/1748#discussion_r270924688> , or mute the thread <https://github.com/notifications/unsubscribe-auth/Aj5tqa8KteA_6tDtOfEcgf0f2BN519ZFks5vciXQgaJpZM4aEW47> .  <https://github.com/notifications/beacon/Aj5tqSOImUv63c8tUvxK4M3J41EuiH4Hks5vciXQgaJpZM4aEW47.gif>

apcameron commented:
@ynezz Rebased and addressed the issues that needed resolving and re-complied and tested on the device

ynezz reviewed (commented):
> Hopefully the needs-changes tag can be removed now.

I've added few more comments, and as already explained in the other comment:

> > ... and your references to #1704 in regards to common DTSI, dynamic partitions etc.

If I put aside that DTSI remark (which could be solved later and I would merge this PR if it was blocked just by this point), I think, that in order to move forward with this PR, we probably need to handle dynamic partitions and MTD splitter as being discussed in #1704 OR we should make a conclusion, that it's perhaps wrong direction and we should simply merge it in the current state.

adrianschmutzler commented:
> Please use tp-link:green:lan to stay consistent with naming conventions established already in that file. Or is there any reason for that lan0 name?

Although slightly off-topic:
This is an obstacle when you really want to merge definitions in DTSI files: If you have two LAN ports, you will name them "lan0" and "lan1", if you have one, you will name it "lan". However, there are cases like for the CPE210-v1 (two ports) and the CPE210-v2 (one port), where it would make sense to put the definition for one LED into the DTSI, and then only add the second one in the device DTS. Then IMHO it would be better to have a "lan0" at the single-port than to have "lan" and "lan1" at the two-port device.

With this background, I would actually consider it preferable to generally use "lan0" for one-port devices, as we are, for example, totally accustomed to for "eth0".

ynezz commented:
> This is an obstacle when you really want to merge definitions in DTSI files:
> If you have two LAN ports, you will name them "lan0" and "lan1", 

No, you'll name them lan1 and lan2 as per already estabilished naming scheme.

> if you have one, you will name it "lan". 

Yeah, exactly my point.

> However, there are cases like for the CPE210-v1 (two ports) and the
> CPE210-v2 (one port), where it would make sense to put the definition for
> one LED into the DTSI, and then only add the second one in the device DTS.
> Then IMHO it would be better to have a "lan0" at the single-port than to
> have "lan" and "lan1" at the two-port device.

As I've explained before, I was bitten many times by this premature upfront
optimizations attempts, so please make a decision once there's such complete
DTS/DTSI case shown on the table. Anyway, in order to stay consistent, the
labels should be lan and lan1,lan2.

apcameron commented:
@ynezz @adrianschmutzler I have updated it as you suggested. As far as Dynamic partitions and reorganization attempts we should probably wait until most of the TP-LINK CPE devices have been ported from the ar71xx branch to better understand what can be optimized and how best to implement Dynamic Partitions. This would be an exercise for someone who understands the big picture.

apcameron commented:
@ynezz Is this PR ready for you to Remove the Needs Changes Tag or are there more changes needed?

apcameron commented:
@ynezz Do I need to do anything to get a reviewer?

apcameron commented:
Is it possible for someone to review this soon so it can be merged?

apcameron commented:
@NeoRaider or @robimarko would either of you be able to review this PR for me as you are famaliar with these devices.
I have it running on my CPE610 V1 and can provide logs if needed.

robimarko commented on target/linux/ath79/dts/ar9344_tplink_cpe610-v1.dts:
```diff
@@ -0,0 +1,141 @@
+				#address-cells = <1>;
+				#size-cells = <1>;
+
+				partition@0 {
```
Why are you nesting another fixed-partitions node here?
You should only have firmware partition and use a compatible firmware splitter.
A splitter for safe loader already exists, and you should use that.

ynezz replied:
https://github.com/openwrt/openwrt/pull/1748#pullrequestreview-222603970

robimarko replied:
Didn't read anything above last push.
Sorry for saying basically what you already said.

adrianschmutzler replied:
partition@40000 {
				label = "firmware";
				reg = <0x040000 0x780000>;
				compatible = "tplink,firmware";
			};

will do it.

apcameron replied:
I tried it and it gets stuck in a bootloop

robimarko commented on tools/firmware-utils/src/tplink-safeloader.c:
```diff
@@ -278,6 +278,45 @@ static struct device_info boards[] = {
+			{"default-mac", 0x30000, 0x00020},
+			{"product-info", 0x31100, 0x00100},
+			{"signature", 0x32000, 0x00400},
+			{"os-image", 0x40000, 0x1c0000},
```
Here you should use dynamic partitions so that it will be autogenerated and adjusted when building.
Example:
https://github.com/openwrt/openwrt/commit/a2a972b2cfd762bc2207ffddde228737245ab926#diff-b232b41a358a2b44c64fb32925bd265a

adrianschmutzler replied:
This PR uses the OKLI-loader, which the linked commit doesn't.
Will the OKLI-loader interfere with dynamic partitioning?
If not, changes should be easy...

robimarko replied:
It should not interfere.
Why is okli even used?

adrianschmutzler replied:
Just put
{"firmware", 0x40000, 0x770000},
instead of os-image and file-system

Do NOT change first_/last_sysupgrade_partition

apcameron replied:
I tried it and it gets stuck in a bootloop

ynezz commented:
@robimarko https://github.com/openwrt/openwrt/pull/1748#pullrequestreview-222603970

apcameron commented:
@robimarko I have a question in that I also plan to add support to the ar71xx branch for this device.
1) What impact would that have on dynamic partitioning? Would the Example you listed above work for both?
2) Should I add all the changes for the ar71xx branch to this PR or create another one once this PR is finalized?

robimarko commented:
1. It works on ar71xx and ath79 in the same way.
2. Yes, it should be a different PR but I don't think that you will get anything new into ar71xx as it's pretty much in bugfix mode and will be dropped after next stable release.
3. It should not, why are you using okli loader anyway?

adrianschmutzler commented on target/linux/ath79/image/generic-tp-link.mk:
```diff
@@ -135,6 +135,20 @@ define Device/tplink_cpe210-v3
+  LOADER_FLASH_OFFS := 0x43000
+  COMPILE := loader-$(1).elf
+  COMPILE/loader-$(1).elf := loader-okli-compile
+  KERNEL := kernel-bin | append-dtb | lzma | uImage lzma -M 0x4f4b4c49 | loader-okli $(1) 12288
```
Here I am not sure whether this will work for dynamic partitioning. WITHOUT okli you find the following two options:
KERNEL := kernel-bin | append-dtb | lzma | tplink-v1-header -O
KERNEL := kernel-bin | append-dtb | lzma | uImageArcher lzma

apcameron replied:
I tried it and it gets stuck in a bootloop

adrianschmutzler replied:
I will soon (maybe on Monday) receive a CPE210v1 and will then give it a try myself.

apcameron commented:
@robimarko in PR #1704 when discussing dynamic partitions and the OLKI Loader  @NeoRaider **said
Well, the commit message of 0c0bb62 basically explains the situation. Old OpenWrt installations with a too small kernel partition ("os-image") in the TP-Link partition table exist in the field, so the bootloader would fail to load the whole compressed kernel into the RAM. The OKLI loader is a workaround for this: The TP-Link bootloader only loads the tiny OKLI loader, and the OKLI loader will then read the actual kernel from the flash.
While this is only an issue for old installations, and installing a new OpenWrt factory image from the stock firmware will result in a "fixed" partition layout (with bigger os-image), 1. we can't easily upgrade old devices (as explained in the referenced commit message, such an upgrade would need to happen before sysupgrade), and 2. the OKLI loader solution is more robust in case we ever need to increase kernel partition size again. Therefore, I would propose to keep the OKLI loader for ath79 as well.******

He recommended that we stay with the OKLI Loader for these devices. He was the one who did the Initial port from the CPE5xx devices.

apcameron commented:
@adrianschmutzler  @ynezz @robimarko I have done some tests with the suggested changes
With tools/firmware-utils/src/tplink-safeloader.c for the device set to
                .partitions = {
                        {"fs-uboot", 0x00000, 0x20000},
                        {"partition-table", 0x20000, 0x02000},
                        {"default-mac", 0x30000, 0x00020},
                        {"product-info", 0x31100, 0x00100},
                        {"signature", 0x32000, 0x00400},
                        **{"firmware", 0x40000, 0x770000},**
                        {"soft-version", 0x7b0000, 0x00100},
                        {"support-list", 0x7b1000, 0x00400},
                        {"user-config", 0x7c0000, 0x10000},
                        {"default-config", 0x7d0000, 0x10000},
                        {"log", 0x7e0000, 0x10000},
                        {"radio", 0x7f0000, 0x10000},
                        {NULL, 0, 0}
                },


And the DTS changed to

                        partition@40000 {
                                label = "firmware";
                                reg = <0x040000 0x780000>;
                                compatible = "tplink,firmware";
                        };

**With OKLI Loader It just boot loops**
With other bootloaders and no OKLI Loader
KERNEL := kernel-bin | append-dtb | lzma | tplink-v1-header -O (**Nothing**)
KERNEL := kernel-bin | append-dtb | lzma | uImageArcher lzma  (**Nothing**)

In PR #1704 @p-wassi also tried unsuccessfully to get the CPE510 to work with dynamic partitions and these devices are almost the same.
**We may have to stay with fixed partitions for these devices until someone finds a way to make dynamic partitions work for the TP-LINK CPE Devices that were configured to use the OKLI Loader.**

@robimarko do you have an older TP-LINK CPE510 v1 or CPE210 v1 that you can experiment with on this?

adrianschmutzler commented:
> KERNEL := kernel-bin | append-dtb | lzma | uImageArcher lzma (Nothing)

This only works with
compatible = "denx,uimage";
in partition@40000

However, I do not understand why it would not work without OKLI. I think the problems in https://github.com/openwrt/openwrt/pull/1704 are also arising from the combination of both?

BTW: I am only suggesting, I do not have any authority about which way you should go ...

apcameron commented:
The way I see it is unless the code for the mtd splitter is changed to support these devices using the OKLI loader we have no option but to stay with the fixed partitions on the older TP Link CPExxx devices

apcameron commented:
@juhosg any comments on updating the mtd splitter

981213 commented:
TP-Link's U-boot needs an ELF header and an ELF header isn't helpful for writing an mtdsplit driver.
Since OKLI loads an LZMA compressed uImage kernel, it's possible to create a mtd partition starting at that uImage header and use uImage split driver directly. (I think I mentioned this somewhere else.)

BTW I think it's fine to keep current fixed partition layout because there isn't an existing solution in current OpenWrt tree and we need to "invent" something for that.

apcameron commented:
Based on the comment above who needs to approve this PR being merged using fixed partitions so it can proceed?

apcameron commented:
@ynezz Updated to resolve file contention and remove OKLI Loader
Latest BootLog is Attached
[dmesg.txt](https://github.com/openwrt/openwrt/files/3316149/dmesg.txt)

adrianschmutzler commented on target/linux/ath79/dts/ar9344_tplink_cpe610-v1.dts:
```diff
@@ -0,0 +1,141 @@
+
+	keys {
+		compatible = "gpio-keys-polled";
+		poll-interval = <20>;
```
If you still want to try something:
It should be possible to use "gpio-keys" binding here and remove poll-interval.

adrianschmutzler commented on target/linux/ath79/dts/ar9344_tplink_cpe610-v1.dts:
```diff
@@ -0,0 +1,141 @@
+			};
+
+			productinfo: partition@30000 {
+				label = "product-info";
```
Please update both alias and label to "info".
I have recently made an effort to harmonize this: 3d153323ab

adrianschmutzler commented on target/linux/ath79/dts/ar9344_tplink_cpe610-v1.dts:
```diff
@@ -0,0 +1,141 @@
+	status = "okay";
+
+	mtd-cal-data = <&art 0x1000>;
+	mtd-mac-address = <&productinfo 0x08>;
```
Update to &info if you follow my request from above.

adrianschmutzler commented on target/linux/ath79/dts/ar9344_tplink_cpe610-v1.dts:
```diff
@@ -0,0 +1,141 @@
+&eth0 {
+	status = "okay";
+
+	mtd-mac-address = <&productinfo 0x08>;
```
Update to &info if you follow my request from above.

apcameron commented:
@adrianschmutzler Requested updates have been made and tested.

chunkeey commented:
Merged. Thanks.

adrianschmutzler commented:
Although this PR has been has been closed already, I would be interested in some additional data about the device(s) in case you still have access to it:

I'm currently working on a PR to store the label-mac-address in the DTS (https://github.com/openwrt/openwrt/pull/2159).

I would appreciate if you could provide information about your device:
Please run
`find /proc/device-tree/ -name "*mac-address*"`
on your device (just with the current firmware, no need to update anything; just has to be ath79, no ar71xx).
For each of the returned paths (if there are any), retrieve the mac-address with `cat`, e.g.
`cat /proc/device-tree/ahb/eth@19000000/mac-address`

Now check which of those returned MAC addresses matches the one on the label/sticker/case of your device and tell me the device-tree path leading to it. Note that you do not have to post your MAC addresses here, but just a statement about which match, e.g. like
```
/proc/device-tree/ahb/eth@1a000000/mac-address
/proc/device-tree/ahb/eth@1a000000/mtd-mac-address-increment
/proc/device-tree/ahb/eth@1a000000/mtd-mac-address
/proc/device-tree/ahb/eth@19000000/mac-address <== this one
/proc/device-tree/ahb/eth@19000000/mtd-mac-address
/proc/device-tree/ahb/apb/wmac@18100000/mac-address
/proc/device-tree/ahb/apb/wmac@18100000/mtd-mac-address
```

There may be one, two or no paths giving the correct address. It may also happen that the `find` command returns nothing. Thanks.

adrianschmutzler commented:
Unfortunately, my description is slightly incorrect. The MAC addresses are stored binary, so you can't read with `cat`, but have to use the following code to read from the paths retrieved by `find`:
```
. /lib/functions/system.sh
get_mac_binary "$yourpath" 0
```
The include is obviously only required once. Sorry.

apcameron commented:
@adrianschmutzler here is the results of your request
/proc/device-tree/ahb/eth@19000000/mac-address <<== This One
/proc/device-tree/ahb/eth@19000000/mtd-mac-address
/proc/device-tree/ahb/apb/wmac@18100000/mac-address <<== This One
/proc/device-tree/ahb/apb/wmac@18100000/mtd-mac-address

adrianschmutzler commented:
Thanks, so I will put `label-mac-device = &eth0;`.
